Resource Provider - S3

youngjeong46 opened the pull request:
*Issue #, if available:*

*Description of changes:* Creating an S3 Resource Provider. Can be used as log storage mechanism for Apache Airflow add-on (in flight), and other addons as necessary.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

elamaran11 commented on lib/resource-providers/s3.ts:
```diff
@@ -0,0 +1,36 @@
+
+    /**
+     * @param s3BucketName name of the S3 Bucket to look up
+     * @param id  optional id for the structure (for tracking). set to s3bucketname by default
```
Extra space after id.

youngjeong46 replied:
addressed

elamaran11 commented on lib/resource-providers/s3.ts:
```diff
@@ -0,0 +1,36 @@
+
+    /**
+     * Creates the S3 provider.
+     * @param name Name of the S3 Bucket. This must be 
```
This must be what ? Sentence is broken

youngjeong46 replied:
addressed

elamaran11 reviewed (commented):
@youngjeong46 Great PR, appreciate you doing this. 
- Overall i would like to understand the motivation behind this S3 Provider and what is the ask? 
- Where is the S3 provider used in the blueprints framework? Are we using this any where in patterns?
- Could we not use ACK to create a controller which basically allows customers to create S3 in GitOps approach which is preferred. CrossPlane Addon is also available.

Open to have a discussion on the motivation/ask/usage. But im good with approach in the PR with the S3Provider

youngjeong46 commented:
> @youngjeong46 Great PR, appreciate you doing this.
> 
> * Overall i would like to understand the motivation behind this S3 Provider and what is the ask?
> * Where is the S3 provider used in the blueprints framework? Are we using this any where in patterns?
> * Could we not use ACK to create a controller which basically allows customers to create S3 in GitOps approach which is preferred. CrossPlane Addon is also available.
> 
> Open to have a discussion on the motivation/ask/usage. But im good with approach in the PR with the S3Provider. Docs for resource provider needs to be updated if we are pursuing this approach.

* Overall i would like to understand the motivation behind this S3 Provider and what is the ask? As part of the self-managed Apache Airflow addon for EKS Blueprints (in-flight), S3 buckets are used as a way for the webserver and worker pods to write remote logs to S3 buckets. 
* Where is the S3 provider used in the blueprints framework? Are we using this any where in patterns? It will be used in the Apache Airflow pattern that will follow the airflow addon PR (in-progress). I have been testing my own pattern successfully as part of the addon PR using this particular provider (which is now its own separate PR).
* Could we not use ACK to create a controller which basically allows customers to create S3 in GitOps approach which is preferred. CrossPlane Addon is also available. The idea here is that the bucket should be created as part of the Airflow platform for the developers/operators **before**. Helm configuration requires custom values to register the S3 bucket and the IRSA required for the worker pods and webserver to be able to access and write to S3 bucket.

elamaran11 commented:
> > @youngjeong46 Great PR, appreciate you doing this.
> > 
> > * Overall i would like to understand the motivation behind this S3 Provider and what is the ask?
> > * Where is the S3 provider used in the blueprints framework? Are we using this any where in patterns?
> > * Could we not use ACK to create a controller which basically allows customers to create S3 in GitOps approach which is preferred. CrossPlane Addon is also available.
> > 
> > Open to have a discussion on the motivation/ask/usage. But im good with approach in the PR with the S3Provider. Docs for resource provider needs to be updated if we are pursuing this approach.
> 
> @elamaran11
> 
> * Overall i would like to understand the motivation behind this S3 Provider and what is the ask? **As part of the self-managed Apache Airflow addon for EKS Blueprints (in-flight), S3 buckets are used as a way for the webserver and worker pods to write remote logs to S3 buckets.**
> * Where is the S3 provider used in the blueprints framework? Are we using this any where in patterns? **It will be used in the Apache Airflow pattern that will follow the airflow addon PR (in-progress). I have been testing my own pattern successfully as part of the addon PR using this particular provider (which is now its own separate PR).**
> * Could we not use ACK to create a controller which basically allows customers to create S3 in GitOps approach which is preferred. CrossPlane Addon is also available. **The idea here is that the bucket should be created as part of the Airflow platform for the developers/operators _before_. Helm configuration requires custom values to register the S3 bucket and the IRSA required for the worker pods and webserver to be able to access and write to S3 bucket.**

Understood your use-case now and the need for S3Provider is very clear and obvious. LGTM.

elamaran11 reviewed (approved):
@youngjeong46 LGTM

shapirov103 commented on lib/spi/types.ts:
```diff
@@ -110,7 +110,8 @@ export enum GlobalResources {
     Certificate = 'certificate',
-    KmsKey = 'kms-key'
+    KmsKey = 'kms-key',
+    S3 = 's3',
```
I am not sure we need S3 in global resources that in fact is pointing to single bucket. it is as if we stating "the bucket for the blueprint" which I don't think there is such a concept, as opposed to "the VPC", "the KMS Key", etc. I assume customers will use bespoke names like `"my-bucket", new CreateS3BucketProvider` and such.

youngjeong46 replied:
addressed

shapirov103 reviewed (commented):
I am sure, it will be very popular. 

How do we test it? I see no indication that it works. 

Also docs: it we need to let customers know about this provider (may be in the Resource Providers doc). Also an example with an SA that can access S3 such as bucket.grantRead(role) would help.

shapirov103 reviewed (changes requested):
Waiting on feedback to be addressed

youngjeong46 commented on lib/resource-providers/s3.ts:
```diff
@@ -0,0 +1,36 @@
+/**
+ * S3 provider that imports S3 bucket into the current stack by name. 
+ */
+export class ImportS3Provider implements spi.ResourceProvider<s3.IBucket> {
```
addressed

youngjeong46 commented on lib/resource-providers/s3.ts:
```diff
@@ -0,0 +1,36 @@
+/**
+ * S3 provider that creates a new S3 Bucket. 
+ */
+export class CreateS3Provider implements spi.ResourceProvider<s3.IBucket> {
```
addressed

shapirov103 commented:
@youngjeong46 


build (18.12.1)Node.js 12 actions are deprecated. Please update the following actions to use Node.js 16: actions/checkout@v2, actions/setup-node@v1, actions/cache@v2. For more information see: https://github.blog/changelog/2022-09-22-github-actions-all-actions-will-begin-running-on-node16-instead-of-node12/.Show more
--
build (18.12.1): test/resource-providers/s3.test.ts#L2'Match' is defined but never used
build (18.12.1): test/resource-providers/s3.test.ts#L44Missing semicolon

[build (18.12.1)](https://github.com/aws-quickstart/cdk-eks-blueprints/actions/runs/5052978072/jobs/9066364061)
Node.js 12 actions are deprecated. Please update the following actions to use Node.js 16: actions/checkout@v2, actions/setup-node@v1, actions/cache@v2. For more information see: https://github.blog/changelog/2022-09-22-github-actions-all-actions-will-begin-running-on-node16-instead-of-node12/.
[build (18.12.1): test/resource-providers/s3.test.ts#L2](https://github.com/aws-quickstart/cdk-eks-blueprints/pull/689/files#annotation_11448496584)
'Match' is defined but never used
[build (18.12.1): test/resource-providers/s3.test.ts#L44](https://github.com/aws-quickstart/cdk-eks-blueprints/pull/689/files#annotation_11448496588)
Missing semicolon

elamaran11 reviewed (commented):
@youngjeong46 Could we also add the S3Provider usage to example stack.

shapirov103 reviewed (approved):
@youngjeong46 I am good with the code. Similar to Ela's comment, I think it is quite harmless to add an s3bucket provider to the e2e test as well, it will be destroyed after each test I assume.

youngjeong46 commented:
I've added it to the `examples` provider. I changed the create provider's props slightly to allow for adding the standard `BucketProps` (i.e. retention policy, as retaining by default could potentially hit an account's S3 bucket limit).

shapirov103 reviewed (approved):
LGTM

shapirov103 reviewed (approved):
LGTM
